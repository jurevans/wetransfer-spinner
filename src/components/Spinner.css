.progress-bar {
  stroke: #479ef2;
  animation-name: rotateProgress;
  animation-fill-mode: forwards;
  animation-timing-function: linear;
  animation-iteration-count: infinite;
}

.progress-circle {
  stroke: #e9e9e9;
}

@keyframes rotateProgress {
  to {
    transform: rotate(360deg);
  }
}

.percentage {
  font-family: Arial, Helvetica, sans-serif;
  fill: #333;
  font-size: 88px;
}

.percentage-sup {
  fill: #c3c3c3;
  font-size: 32px;
  font-weight: bold;
}
