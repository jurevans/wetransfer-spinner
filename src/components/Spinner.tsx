import { CSSProperties, FC, ReactElement } from 'react';
import { getOffset, getValidPercentage } from '../util';
import './Spinner.css';

type Props = {
  progress?: number;
  animationDuration?: number;
  size?: number | string;
  radius?: number,
  strokeWidth?: number,
  progressStyles?: CSSProperties,
  rotationFactor?: number,
};

const Spinner: FC<Props> = (props: Props): ReactElement => {
  const {
    progress = 0,
    animationDuration = 1,
    size = '100%',
    strokeWidth = 25,
    progressStyles = {},
    rotationFactor = 1,
  } = props;

  const radius = 175;
  const diameter = Math.round(Math.PI * radius * 2);

  const progressStyle: CSSProperties = {
    ...progressStyles,
    strokeDashoffset: getOffset(progress, diameter),
    transition: `stroke-dashoffset ${animationDuration}s ease-out`,
  };

  const validPercentage = getValidPercentage(progress);

  return (
    <svg width={size} height={size} viewBox='-25 -25 400 400'>
      <circle
        className='progress-circle'
        cx={radius}
        cy={radius}
        r={radius}
        strokeWidth={strokeWidth}
        fill='none' />
      <circle
        className='progress-bar'
        cx={radius}
        cy={radius}
        r={radius}
        transform={`rotate(${validPercentage * rotationFactor} ${radius} ${radius})`}
        strokeWidth={strokeWidth}
        strokeDasharray='1100'
        strokeDashoffset='1100'
        strokeLinecap='round'
        style={progressStyle}
        fill='none'
      />
      <text
        className='percentage'
        x={radius}
        y={radius}
        textAnchor='middle'
        dominantBaseline='central'
      >
        <tspan className='percentage-value'>{validPercentage}</tspan>
        <tspan dx={6} dy={-14} className='percentage-sup'>%</tspan>
      </text>
    </svg>
  );
};

export default Spinner;
