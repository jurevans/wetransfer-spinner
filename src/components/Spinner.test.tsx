import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import Spinner from './Spinner';

describe('Spinner', () => {
  it('Renders correctly', () => {
    const tree = renderer.create(<Spinner progress={55} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should initialize at zero with no props provided', () => {
    const { container } = render(<Spinner />);
    // This type of test would be better suited to Enzyme:
    expect(container.getElementsByClassName('percentage-value')[0].innerHTML).toBe('0');
  });

  it('should set the percentage value equal to the prop progress', () => {
    const { container } = render(<Spinner progress={35} />);
    // This type of test would be better suited to Enzyme:
    expect(container.getElementsByClassName('percentage-value')[0].innerHTML).toBe('35');
  });
});
