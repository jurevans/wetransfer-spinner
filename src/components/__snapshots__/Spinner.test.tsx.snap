// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Renders correctly 1`] = `
<svg
  className="progress-spinner"
  height="100%"
  viewBox="-25 -25 400 400"
  width="100%"
>
  <circle
    className="progress-circle"
    cx={175}
    cy={175}
    fill="none"
    r={175}
    strokeWidth={25}
  />
  <circle
    className="progress-bar"
    cx={175}
    cy={175}
    fill="none"
    r={175}
    strokeDasharray="1100"
    strokeDashoffset="1100"
    strokeLinecap="round"
    strokeWidth={25}
    style={
      Object {
        "animationDuration": "2s",
        "animationPlayState": "paused",
        "strokeDashoffset": 495,
        "transformOrigin": "175px 175px",
        "transition": "stroke-dashoffset 1s ease-out",
      }
    }
    transform="rotate(-90 175 175)"
  />
  <text
    className="percentage"
    dominantBaseline="central"
    textAnchor="middle"
    x={175}
    y={175}
  >
    <tspan
      className="percentage-value"
    >
      55
    </tspan>
    <tspan
      className="percentage-sup"
      dx={6}
      dy={-14}
    >
      %
    </tspan>
  </text>
</svg>
`;
